### PR TITLE
Feature/check for equal engines

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

**Comment by jsteemann** [comment]
This will not work at all.

**Comment by jsteemann** [comment]
The feature has "Logger" as its only dependency. So it may start directly after the Logger, when ClusterComm and all this is completely dysfunctional. It **must** start after the coordinator is up and running and communication to other cluster nodes has been established.

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    requests.emplace_back(id, rest::RequestType::POST, url, bodyToSend);
+  }
+
+  // send requests
+  std::size_t requestsDone = 0;
+  cc->performRequests(requests, 60.0 /*double timeout*/, requestsDone
+                    ,LogTopic("Storage Engine Equality EngineEqualityCheckFeature")
+                    ,false);
+
+  // FIXME make sure all servers responed or fail
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: who is going to fix this?

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  // check answers
+  for (auto const& request : requests){
+    auto const& result = request.result;
+    if(result.status == CL_COMM_RECEIVED) {
+      httpclient::SimpleHttpResult const& simpleResult = *request.result.result;
+      std::string body = simpleResult.getBody().c_str();
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: why convert `std::string&` to `char const*` to `std::string` here?

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+      //FIXME extract engine from body
+      LOG_DEVEL << body;
+      auto dbserverEngine = engineName;
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: mmh. how can the engines be different here *at all*?

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  // FIXME make sure all servers responed or fail
+
+  // check answers
+  for (auto const& request : requests){
+    auto const& result = request.result;
+    if(result.status == CL_COMM_RECEIVED) {
+      httpclient::SimpleHttpResult const& simpleResult = *request.result.result;
+      std::string body = simpleResult.getBody().c_str();
+
+      //FIXME extract engine from body
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: yes

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  auto bodyToSend = std::make_shared<std::string>();
+  std::string const url = "/_api/version";
+  for(auto const& id : serverIdVector){
+    requests.emplace_back(id, rest::RequestType::POST, url, bodyToSend);
+  }
+
+  // send requests
+  std::size_t requestsDone = 0;
+  cc->performRequests(requests, 60.0 /*double timeout*/, requestsDone
+                    ,LogTopic("Storage Engine Equality EngineEqualityCheckFeature")
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: why add a one-time LogTopic here?

**Comment by ObiWahn** [comment]
This is WIP it was not ready for review. According to (the github workflow document) I am not allowed to set labels form group 9.

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp (hunk 2 above)
Comment (ObiWahn) [reply lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: The body is a arangodb::basics::StringBuffer or something other non-standard. I expected the body of a version to be short and wanted to have it as string for easy inspection.

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp (hunk 1 above)
Comment (ObiWahn) [reply lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: I'll do

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp (hunk 3 above)
Comment (ObiWahn) [reply lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: They will never be different.

**Comment by jsteemann** [comment]
Ok. Just let us know when this PR is ready.

File: lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+using namespace arangodb;
+
+EngineEqualityCheckFeature::EngineEqualityCheckFeature(
+    application_features::ApplicationServer* server)
+    : ApplicationFeature(server, "EngineEqualityCheck") {
+  setOptional(false);
+  requiresElevatedPrivileges(false);
+  startsAfter("Cluster");
+}
```
Comment (jsteemann) [inline lib/ApplicationFeatures/EngineEqualityCheckFeature.cpp]: Is this late enough? do we also depend on other features?
I think the new feature should even start after the ServerFeature, probably even after the BootstrapFeature.

File: arangod/RestServer/arangod.cpp [hunk 7]
```diff
@@ -129,7 +130,7 @@ static int runServer(int argc, char** argv, ArangoGlobalContext &context) {
     std::vector<std::string> nonServerFeatures = {
         "Action",        "Agency",
         "Cluster",       "Daemon",
-        "FoxxQueues",    "GeneralServer", 
+        "FoxxQueues",    "GeneralServer",
```
Comment (jsteemann) [inline arangod/RestServer/arangod.cpp]: add the new feature here too?

**Comment by jsteemann** [comment]
Is there a Jenkins run for this PR? If not, please start one.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/171/

**Pull request merged**